zebra/zebra_pw.c

```diff
@@ -142,18 +142,6 @@ check_lsp (struct zebra_pw_t *pw)
       zlog_warn ("No rib found for PW %u at VPN %s", pw->pwid, pw->vpn_name);
       return 1;
     }
-  /* check labels for each nexthop in Route */
-  /*
-   * Need to ensure that there's a label binding for all nexthops.
-   * Otherwise, ECMP for this route could render the pseudowire unusable.
-   */
-  for (ALL_NEXTHOPS_RO(rib->nexthop, nexthop, tnexthop, recursing))
-    if (!nexthop->nh_label || nexthop->nh_label->num_labels == 0)
-      {
-        zlog_warn ("No label found in rib for PW %u at VPN %s", pw->pwid,
-                   pw->vpn_name);
-        return 1;
-      }
```

Comment:
> If the label is implicit null it's not installed in Zebra, so this check doesn't make sense.

Not really, you can't remove this.

The P from VPLS stands for Private. Removing these lines introduces a security issue as you break the isolation between the VPLS customers.

Receiving an implicit-null label for the pseudowire's remote endpoint is different than receiving no label at all.

If you receive an implicit-null label, it means that the remote endpoint is directly connected to you. In this case, due to the default PHP behavior, you can send the pseudowire frames with the VPN label only. You don't need to push an additional label to tunnel the packets to the remote endpoint.

Now let's say that you didn't receive any label for the pseudowire's remote endpoint. In this case, you don't know if the remote endpoint is directly connected to you or not. If you send L2VPN frames pushing just one label (the VPN label) and the nexthop is not the remote endpoint, then it will not understand what to do with the VPN label. Even worse, it might have allocated the very same label for another VPLS client and forward the packet to it. Now you have a security issue. This is possible because labels are locally allocated in standard MPLS networks (when segment routing is not in use).

With that said, the real problem here is that we're not adding LDP's implicit-null labels to routes. Removing the `out_label != MPLS_IMP_NULL_LABEL` checks below (zserv.c) should solve the problem.

```
  if (command == ZEBRA_MPLS_LABELS_ADD)
    {
      mpls_lsp_install (zvrf, type, in_label, out_label, gtype, &gate,
                        ifindex);
      if (out_label != MPLS_IMP_NULL_LABEL)
        mpls_ftn_update (1, zvrf, type, &prefix, gtype, &gate, ifindex,
                         distance, out_label);
    }
  else if (command == ZEBRA_MPLS_LABELS_DELETE)
    {
      mpls_lsp_uninstall (zvrf, type, in_label, gtype, &gate, ifindex);
      if (out_label != MPLS_IMP_NULL_LABEL)
        mpls_ftn_update (0, zvrf, type, &prefix, gtype, &gate, ifindex,
                         distance, out_label);
    }
```